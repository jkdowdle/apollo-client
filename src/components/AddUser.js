import React from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import { usersFeedQuery, UsersQuery } from './AllUsers'

const AddUserMutation = gql`
  mutation AddUser($input: UserInput!) {
    addUser(input: $input) {
      id
      firstName
      age
    }
  }
`

export class AddUser extends React.Component {
  input = {
    firstName: '',
    age: ''
  }

  updateForm = event => {
    const target = event.target
    const name = target.name
    const value = target.value

    this.input[name] = value
  }

  render() {
    const props = this.props
    return (
      <div>
        <h2>Add User</h2>
        <input type="text" name="firstName" onChange={this.updateForm} />
        <br />
        <br />
        {<input type="number" name="age" onChange={this.updateForm} />}
        <br />
        <br />
        <button onClick={props.addUser.bind(null, this.input)}>Add</button>
      </div>
    )
  }
}

const update = (store, { data: { addUser } }) => {
  const data = store.readQuery({ query: UsersQuery, variables: { cursor: null } })
  data.users.push(addUser)
  store.writeQuery({ query: UsersQuery, data })

  // const data = store.readQuery({ query: UsersQuery, variables: { cursor: null } })
  // data.users.unshift(addUser)

  // store.writeQuery({ query: UsersQuery, data })
}

const optimisticResponse = (input) => ({
    __typename: 'Mutation',
    addUser: {
      __typename: 'User',
      id: -1,
      optimistic: true,
      ...input
    }
})

const props = ({ mutate }) => ({
  addUser: (input) => mutate({ 
    variables: { input }, 
    update, 
    optimisticResponse: optimisticResponse(input) 
  })
})

export default graphql(AddUserMutation, {
  props
})(AddUser)
