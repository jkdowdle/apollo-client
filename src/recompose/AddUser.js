import React from 'react'
import { withState, withHandlers, pure, compose } from 'recompose'
import gql from 'graphql-tag'
import { graphql } from 'react-apollo'

import { usersList } from './AllUsers'

const defaultState = { firstName: '', age: '' }

const form = withState('form', 'setForm', defaultState)
const updateForm = ({ form, setForm }) => ({ target: { name, value }}) => setForm({ ...form, [name]: value })

const addUser = ({ addUser, form, setForm }) => (event) => {
  event.preventDefault()
  return addUser({ ...form })
    .then((res) => setForm({ firstName: '', age: '' }))
}

const handlers = withHandlers({
  updateForm,
  addUser
})

const AddUserPure = ({ updateForm, form: { firstName, age }, addUser, ...props }) => {
  return (
    <form onSubmit={addUser}>
      <h2>New User</h2>

      <input
        placeholder="User's name"
        type="text"
        name="firstName"
        value={firstName}
        onChange={updateForm}
      />

      <br />
      <br />

      <input
        type="number"
        name="age"
        value={age}
        onChange={updateForm}
      />

      <button type="submit">Go</button> 

    </form>
  )
}

const addUserMutation = gql`
  mutation AddUser($input: UserInput!) {
    addUser(input: $input) {
      id
      firstName
      age
    }
  }
`

const update = (store, { data: { addUser } }) => {
  const data = store.readQuery({ query: usersList, variables: {} })
  data.users.unshift(addUser)
  store.writeQuery({ query: usersList, data })
}

const optimisticResponse = (input) => ({
  __typename: 'Mutation',
  addUser: {
    __typename: 'User',
    id: -1,
    optimistic: true,
    ...input
  }
})

const props = ({ mutate }) => ({
  addUser: (input) => mutate({ 
    variables: { input }, 
    update, 
    optimisticResponse: optimisticResponse(input) 
  })
})

const mutation = graphql(
  addUserMutation,
  { props }
)

const AddUser = compose(
  mutation,
  form,
  handlers,
  pure
)(AddUserPure)

export default AddUser