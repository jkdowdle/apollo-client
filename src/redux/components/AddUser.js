import React from 'react'

export const AddUser = ({ handleSubmit, handleChange, firstName, age }) => {
  return (
    <form onSubmit={handleSubmit}>
      <div>
        <input type="text" value={firstName} name="firstName" onChange={handleChange} />
      </div>
      <div>
        <input type="text" value={age} name="age" onChange={handleChange} />
      </div>
      <div>
        <button type="submit">AddUSer</button>
      </div>
    </form>
  )
} 

export default AddUser