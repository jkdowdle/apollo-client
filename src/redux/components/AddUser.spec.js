import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import AddUser from './AddUser'

describe('AddUser', () => {
  describe('render', () => {
    let props

    beforeEach(() => {
      props = {
        firstName: 'foo bar',
        age: 1000,
        handleChange: jest.fn(),
        handleSubmit: jest.fn()
      }
    })

    it('should match a default snapshot', () => {
      const wrapper = shallow(<AddUser />)

      expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('should match a default snapshot', () => {
      const wrapper = shallow(<AddUser {...props} />)

      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    let wrapper, props

    beforeEach(() => {
      props = {
        firstName: 'foo bar',
        age: 1000,
        handleChange: jest.fn(),
        handleSubmit: jest.fn()
      }
      wrapper = shallow(<AddUser {...props} />)
    })

    afterEach(() => {
      props.handleChange.mockReset()
      props.handleSubmit.mockReset()
    })

    it('should submit the form with value of the submit event', () => {
      const form = wrapper.find('form')
      form.simulate('submit')

      expect(props.handleSubmit).toHaveBeenCalled()
    })

    it('should call handleSubmit each time any input is changed', () => {
      const nameInput = wrapper.find('input[name="firstName"]')
      const ageInput = wrapper.find('input[name="age"]')

      nameInput.simulate('change')
      nameInput.simulate('change')
      ageInput.simulate('change')
      
      expect(props.handleChange).toHaveBeenCalledTimes(3)
    })
  })
})