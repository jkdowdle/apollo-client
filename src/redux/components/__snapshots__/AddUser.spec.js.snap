// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUser render should match a default snapshot 1`] = `
<form>
  <div>
    <input
      name="firstName"
      type="text"
    />
  </div>
  <div>
    <input
      name="age"
      type="text"
    />
  </div>
  <div>
    <button
      type="submit"
    >
      AddUSer
    </button>
  </div>
</form>
`;

exports[`AddUser render should match a default snapshot 2`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      name="firstName"
      onChange={[Function]}
      type="text"
      value="foo bar"
    />
  </div>
  <div>
    <input
      name="age"
      onChange={[Function]}
      type="text"
      value={1000}
    />
  </div>
  <div>
    <button
      type="submit"
    >
      AddUSer
    </button>
  </div>
</form>
`;
